plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/Stablecoin.hs

```diff
@@ -0,0 +1,385 @@
+    deriving stock (Generic, Haskell.Eq, Haskell.Show)
+    deriving anyclass (ToJSON, FromJSON)
+    deriving Functor via Identity
+
+-- | An amount of base currency coins (eg. Ada or some native currency)
+newtype BC a = BC { unBC :: a }
+    deriving newtype (Haskell.Num, Eq, Ord, AdditiveGroup, AdditiveMonoid, AdditiveSemigroup, MultiplicativeSemigroup)
+    deriving stock (Generic, Haskell.Eq, Haskell.Show)
+    deriving anyclass (ToJSON, FromJSON)
+    deriving Functor via Identity
+
+-- | An amount of peg currency (eg. USD)
+newtype PC a = PC { unPC :: a }
+    deriving newtype (Haskell.Num, Eq, Ord, AdditiveGroup, AdditiveMonoid, AdditiveSemigroup, MultiplicativeSemigroup)
+    deriving stock (Generic, Haskell.Eq, Haskell.Show)
+    deriving anyclass (ToJSON, FromJSON)
+    deriving Functor via Identity
+
+-- | The bank's state
+data BankState =
+    BankState
+        { bsReserves            :: BC Integer -- ^ Value of the bank's reserves in base currency
+        , bsStablecoins         :: SC Integer -- ^ Amount of stablecoins in circulation
+        , bsReservecoins        :: RC Integer -- ^ Amount of reservecoins currently in circulation
+        , bsForgingPolicyScript :: MonetaryPolicyHash -- ^ Hash of the forging policy that forwards all check to the state machine. (This has to be in this type, rather than in 'Stablecoin', to avoid a circular dependency on the script's hash)
+        }
+    deriving stock (Generic, Haskell.Eq, Haskell.Show)
+    deriving anyclass (ToJSON, FromJSON)
+
+-- | Initialise the 'BankState' with zero deposits.
+initialState :: StateMachineClient BankState Input -> BankState
+initialState StateMachineClient{scInstance=StateMachineInstance{validatorInstance}} =
+    BankState
+        { bsReserves = 0
+        , bsStablecoins = 0
+        , bsReservecoins = 0
+        , bsForgingPolicyScript = monetaryPolicyHash $ forwardingMPS $ scriptHash validatorInstance
+        }
+
+{-# INLINEABLE convert #-}
+-- | Convert peg currency units to base currency units using the
+--   observed conversion rate
+convert :: ConversionRate -> PC (Ratio Integer) -> BC (Ratio Integer)
+convert rate (PC pc) =
+    BC $ rate * pc
+
+{-# INLINEABLE liabilities #-}
+-- | The bank's liabilities (total value of stablecoins in base currency)
+liabilities ::
+    BankState
+    -> ConversionRate
+    -> BC (Ratio Integer)
+liabilities BankState{bsReserves=BC reserves,bsStablecoins=SC stablecoins} cr =
+    let BC stableCoinLiabilities = convert cr (PC $ fromInteger stablecoins)
+    in BC (min (fromInteger reserves) stableCoinLiabilities)
+
+{-# INLINEABLE equity #-}
+-- | The bank's equity (what's left of the reserves after subtracting
+--   liabilities).
+equity ::
+    BankState
+    -> ConversionRate
+    -> BC (Ratio Integer)
+equity r@BankState{bsReserves=BC reserves} cr =
+    let BC l = liabilities r cr
+    in BC (fromInteger reserves - l)
```

Comment:
Perhaps it's so this never goes negative? But arguably it *should* be negative if the bank's liabilities exceed the reserves...

Reply:
Yeah I think that's the reason, but I agree that it should be negative if the price goes below a certain level